### PR TITLE
Updates the Task Developer Guide

### LINKED ISSUE #2322: Revisit Task developer guide
As a developer, I'd like to revise and cleanup Task developer guide, so it is up-to-date and consistent with what we do in Stream counterpart.

**Acceptance:**
- Revise the developer guide for correctness
- Include Spring Initializr workflow
- Further clarify DB requirement
- Make it easy to repeat

### PR DESCRIPTION
**Pull request by cppwfs**
resolves #2322

### CODE REVIEW

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc [hunk 1]
```diff
@@ -57,118 +57,233 @@ If you have debug mode turned on, you can get even more information, in a line s
...
 
-. From the command line, run `cd labs/lab2`.
-. From the command line, run `mvn clean package`.
-+
-You should see a typical set of Maven build output for a Spring Boot application.
-. From the command line, run  `java -jar target/tasklab-0.0.1-SNAPSHOT.jar`.
-+
-You should see a Spring application start, print "`Hello World`" to the console, and stop.
+[[task-dev-guide-creating-project]]
+=== Creating the Spring Task Project using Spring Initializer
```
Comment (sabbyanandan) [inline spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: Typo: "Initializer" -> "initializr".

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc [hunk 2]
```diff
@@ -57,118 +57,233 @@ If you have debug mode turned on, you can get even more information, in a line s
...
+Now let's create and test an application that prints a timestamp to the console.
```
Comment (sabbyanandan) [inline spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: Maybe: "timestamp" -> "current time"

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc [hunk 3]
```diff
@@ -57,118 +57,233 @@ If you have debug mode turned on, you can get even more information, in a line s
...
 
-[[task-dev-guide-building-timestamp-task-setting-up-environment]]
-=== Setting up the Environment
+[[task-dev-guide-building-timestamp-task-from-initializer]]
+== Developing Your Timestamp Application
 
-Before we can do anything else, we have to set up our work environment. First, we must install RabbitMQ, as described https://github.com/cppwfs/DNDataflow/blob/master/labs/InstallRabbitnMQ.pdf[in this document].
-Then we need to create a directory and add a starter project (which we created to simplify this setup). To do so:
+So now that we've used a pre-built timestamp task application, let's create one of our own.
+Most IDEs have good support for Apache Maven, so we use it as the build tool for this project.
```
Comment (mminella) [inline spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: Without looking at the stream section of the docs...is that also Maven centric or does it have Gradle instructions as well?

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc [hunk 4]
```diff
@@ -57,118 +57,233 @@ If you have debug mode turned on, you can get even more information, in a line s
...
-=== Getting Started
+[source]
+$ java -version
+java version "1.8.0_31"
+Java(TM) SE Runtime Environment (build 1.8.0_31-b13)
+Java HotSpot(TM) 64-Bit Server VM (build 25.31-b07, mixed mode)
 
-To get started on making a task, we must first make a Spring Boot application. The DNDataflow project that we cloned in the preceding section took care of that for us, but we need to build and run it, as follows:
+[source]
+$ mvn -v
```
Comment (mminella) [inline spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: You shouldn't need Maven installed since a project from Initializr comes with the wrapper...

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc [hunk 5]
```diff
@@ -57,118 +57,233 @@ If you have debug mode turned on, you can get even more information, in a line s
...
+
+	public static void main(String[] args) {
+		SpringApplication.run(TimestampApplication.class, args);
+	}
+
+	public static class TimestampCommandLineRunner implements CommandLineRunner {
+
+		@Override
+		public void run(String... strings) throws Exception {
+			DateFormat dateFormat = new SimpleDateFormat("YYYY-MM-dd");
```
Comment (sabbyanandan) [inline spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: So, we aren't printing the time with this format. Maybe include `HH:ss` also in it, so the Application does produce the time as well?

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc [hunk 6]
```diff
@@ -57,118 +57,233 @@ If you have debug mode turned on, you can get even more information, in a line s
...
-. From the command line, run  `java -jar target/tasklab-0.0.1-SNAPSHOT.jar`.
-+
-You should see a Spring application start, print "`Hello World`" to the console, and stop.
+[[task-dev-guide-creating-project]]
+=== Creating the Spring Task Project using Spring Initializer
+Now let's create and test an application that prints a timestamp to the console.
 
-[[task-dev-guide-building-timestamp-task-making-boot-task]]
-=== Making a Spring Boot Application into a Spring Cloud Task
+To do so:
```
Comment (mminella) [inline spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: We may want to make a note that you can use IntelliJ or STS to do this as well (we just aren't using it here).

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc [hunk 7]
```diff
@@ -57,118 +57,233 @@ If you have debug mode turned on, you can get even more information, in a line s
...
-. Open `src/main/java/io/spring/TasklabApplication.java` in either an IDE or a text editor.
-. On Line 13, add the `@EnableTask` annotation.
-+
-NOTE: If you do not use an IDE, you also need to un-comment lines 6 through 9.
-. Save your changes to `TasklabApplication.java`.
-. Open `src/main/resources/application.properties`.
-. Add the following properties:
-+
+
+While it may seem small, quite a bit is going on. For more about Spring
```
Comment (sabbyanandan) [inline spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: > While it may seem small, quite a bit is going on.

Seems open-ended. We could put a few lines on what is happening or maybe link to SCT docs on this matter?

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc [hunk 8]
```diff
@@ -57,118 +57,233 @@ If you have debug mode turned on, you can get even more information, in a line s
...
+Boot specifics, see the
+http://docs.spring.io/spring-boot/docs/current/reference/html/[Spring Boot reference documentation].
+
+Now let's open the `application.properties` file in `src/main/resources`. We need
```
Comment (sabbyanandan) [inline spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: Possible reduce "We need"'s from this? Repeats.

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc (hunk 5 above)
Comment (mminella) [inline spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: Not exactly a timestamp... ;)

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc [hunk 9]
```diff
@@ -57,118 +57,233 @@ If you have debug mode turned on, you can get even more information, in a line s
...
+spring.application.name=timestamp
 ----
-+
-Setting the logging level for the task to `DEBUG` creates output to let us know that the task works.
-Setting the application name lets us store a meaningful name in the database, so that we can find our task in the `TASK_EXECUTION` table.
-. Save your change to `application.properties`.
 
-Now that we have made our changes, we naturally want to run it and see it work. To do so:
+[[getting-started-at-task]]
+==== The @EnableTask annotation
```
Comment (sabbyanandan) [inline spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: Oh, cool, you do explain here. Maybe it'd be good to remove "While it may seem small, quite a bit is going on." from the previous section.

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc [hunk 10]
```diff
@@ -57,118 +57,233 @@ If you have debug mode turned on, you can get even more information, in a line s
...
-. Add the following properties:
-+
+
+While it may seem small, quite a bit is going on. For more about Spring
+Boot specifics, see the
+http://docs.spring.io/spring-boot/docs/current/reference/html/[Spring Boot reference documentation].
+
+Now let's open the `application.properties` file in `src/main/resources`. We need
+to configure two properties in `application.properties`: We need to set the application
+name (which is translated to the task name), and we need to set the logging for Spring
```
Comment (mminella) [inline spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: Can we reword this to "we want to set the logging"?  I don't want people to think that you _need_ debug logging on for SCT.

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc [hunk 11]
```diff
@@ -57,118 +57,233 @@ If you have debug mode turned on, you can get even more information, in a line s
...
+class-level annotation tells Spring Cloud Task to bootstrap it's functionality. By
+default, it imports an additional configuration class (`SimpleTaskConfiguration`). This
+additional configuration registers the `TaskRepository` and the infrastructure for its
+use.
+
+In our demo, the `TaskRepository` uses an embedded H2 database to record the results
+of a task. This H2 embedded database is not a practical solution for a production environment, since
+the H2 DB goes away once the task ends. However, for a quick getting-started
+experience, we will use this in our example as well as echoing to the logs what is being updated
+in that repository. In the <<features-configuration>> section (later in this
```
Comment (mminella) [inline spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: We probably can leave out the "Later in this documentation" if we're providing a link to it.

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc [hunk 12]
```diff
@@ -57,118 +57,233 @@ If you have debug mode turned on, you can get even more information, in a line s
...
+
+NOTE: Any processing bootstrapped from mechanisms other than a `CommandLineRunner` or
+`ApplicationRunner` (by using `InitializingBean#afterPropertiesSet` for example) is not
+ recorded by Spring Cloud Task.
+
+[[getting-started-running-the-example]]
+=== Running the Example
+
+At this point, our application should work.  Since this application is Spring Boot-based,
+we can run it from the command line by using `$ mvn spring-boot:run` from the root
```
Comment (sabbyanandan) [inline spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: Perhaps remove `$`.

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc [hunk 13]
```diff
@@ -57,118 +57,233 @@ If you have debug mode turned on, you can get even more information, in a line s
...
+When our sample application runs, Spring Boot launches our `TimestampCommandLineRunner`
+and outputs our timestamp message to standard out. The `TaskLifecycleListener`
+records the start of the task and the end of the task in the repository.
+
+[[getting-started-main-method]]
+==== The main method
+
+The main method serves as the entry point to any java application.  Our main method
+delegates to Spring Boot's `SpringApplication` class.  You can read more about it in the
+Spring Boot documentation.
```
Comment (mminella) [inline spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: Link?

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc [hunk 14]
```diff
@@ -57,118 +57,233 @@ If you have debug mode turned on, you can get even more information, in a line s
...
+
+[[getting-started-clr]]
+==== The CommandLineRunner
+
+Spring includes many ways to bootstrap an application's logic. Spring Boot provides
+a convenient method of doing so in an organized manner through its `*Runner` interfaces
+(`CommandLineRunner` or `ApplicationRunner`). A well behaved task can bootstrap any
+logic by using one of these two runners.
```
Comment (mminella) [inline spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: "A well behaved task _will_ bootstrap..."

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc [hunk 15]
```diff
@@ -57,118 +57,233 @@ If you have debug mode turned on, you can get even more information, in a line s
...
+
+  .   ____          _            __ _ _
+ /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
+( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
+ \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
+  '  |____| .__|_| |_|_| |_\__, | / / / /
+ =========|_|==============|___/=/_/_/_/
+ :: Spring Boot ::        (v2.0.3.RELEASE)
+
+2018-07-25 11:11:16.695  INFO 5342 --- [           main] i.s.demo.timestamp.TimestampApplication  : Starting TimestampApplication on Glenns-MacBook-Pro-2.local with PID 5342 (/Users/glennrenfro/project/crap/timestamp/target/classes started by glennrenfro in /Users/glennrenfro/project/crap/timestamp)
```
Comment (mminella) [inline spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: May not want "crap" in your path in the docs...

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc [hunk 16]
```diff
@@ -57,118 +57,233 @@ If you have debug mode turned on, you can get even more information, in a line s
...
 
-. Open `src/main/java/io/spring/TasklabApplication.java` in either an IDE or a text editor.
-. Copy the following line into the file at line 26:
+. Open `src/main/java/io/spring/Timestampapplic.java` in either an IDE or a text editor.
+. Insert the following line into the file at line 32:
 +
 [source,java]
 throw new IllegalStateException("No Task For You!");
-. From the command line, run `mvn clean package -DSkipTests`.
+. From the command line, run `mvn clean spring-boot:run -DskipTests`.
```
Comment (mminella) [inline spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: We should encourage users to use the wrapper.

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc [hunk 17]
```diff
@@ -182,8 +297,7 @@ public void afterTask(TaskExecution taskExecution) {
   System. out.println("After TASK");
 }
 ----
-. From the command line, run `mvn clean package`.
-. From the command line, run `java -jar target/tasklab-0.0.1-SNAPSHOT.jar`.
+. From the command line, run `mvn clean spring-boot:run`.
```
Comment (mminella) [inline spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: Please use the wrapper.

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc [hunk 18]
```diff
@@ -204,40 +318,30 @@ To do so:
   <artifactId>mariadb-java-client</artifactId>
 </dependency>
 ----
-. From the command line, run `mvn clean package`.
-. From the command line, run `java -jar target/tasklab-0.0.1-SNAPSHOT.jar`.
+. From your command line set up the database connection properties for MySql for example
```
Comment (mminella) [inline spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: We should either say we're using Maria or use the MySql URL in the below config.

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc (hunk 3 above)
Comment (markpollack) [reply spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: There is no need to have any content about maven or grade, initialzr can create maven or gradle, up to the user.  Any IDE will successfully import either a gradle or maven project.  We can show the user wants.  Also, I think it is filler to have someone check that they have installed java and show `java -version`.

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc (hunk 9 above)
Comment (cppwfs) [reply spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: Removed that section.

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc (hunk 15 above)
Comment (cppwfs) [reply spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: Good catch :-D

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc (hunk 18 above)
Comment (cppwfs) [reply spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: updated to mysql

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc (hunk 3 above)
Comment (cppwfs) [reply spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: section was removed.

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc (hunk 12 above)
Comment (markpollack) [reply spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: adding ./mvnw consistently...

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc (hunk 15 above)
Comment (markpollack) [reply spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: removed

File: spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc (hunk 18 above)
Comment (markpollack) [reply spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: does it apply to both?

Comment (mminella) [reply spring-cloud-dataflow-docs/src/main/asciidoc/task-developer-guide.adoc]: While it does, I'd just like us to be consistent.  If we say in the docs we're using MySql...use MySql, etc.

**Comment by markpollack** [comment]
merged as 5976bef7404ea1bcbc860817b02423fd934c2548
